### PR TITLE
refactor: improved loading algorithm of options and configs

### PR DESCRIPTION
**Pull request by evilebottnawi**
Why?
1. No todo :smile:
2. Always load config if your pass `config.path`, now it is not working as expected if i pass `config.path` and `sourceMap: false` my config file not loading -> `options.config && !sourceMap && length > 1` (i want to load config and set `sourceMap` use loader options ), also sometimes i want setup some option from options loader.
3. Some perf (we don't create `rc` object because it is not necessary if i use only `options` without `options.config`)
4. Allow to passing options to `postcss` throw `loader.options`, now it is possibly only using `postcss` config.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -47,78 +47,67 @@ module.exports = function loader (css, map) {
...
-    }
 
     return postcssrc(rc.ctx, rc.path, { argv: false })
   }).then((config) => {
     if (!config) config = {}
 
     if (config.file) this.addDependency(config.file)
 
+    const sourceMap = options.sourceMap
+    delete options.sourceMap
```
Comment (michael-ciniawsky) [inline lib/index.js]: This caused problems on 'multiple' runs for some weird reason, but maybe because it was defined in the wrong place, not sure here tbh ¯\_(ツ)_/¯. It work for the first loader run and was then missing for all further runs, which lead to `const sourceMap` being always `undefined`. Besides that `delete options.sourceMap` => `options.sourceMap = null` (perf) ?

File: lib/index.js [hunk 2]
```diff
@@ -47,78 +47,67 @@ module.exports = function loader (css, map) {
...
-    path: path.dirname(file),
-    ctx: {
-      file: {
-        extname: path.extname(file),
-        dirname: path.dirname(file),
-        basename: path.basename(file)
-      },
-      options: {}
+  Promise.resolve().then(() => {
+    if (!options.config || (options.config && !options.config.path)) {
```
Comment (michael-ciniawsky) [inline lib/index.js]: This changes to config loading behaviour ? See comment below where `options` get reassigned

File: package.json [hunk 3]
```diff
@@ -33,6 +33,7 @@
     "docs": "jsdoc2md lib/index.js > LOADER.md",
     "pretest": "npm run lint && npm run test:build",
     "test": "jest",
+    "test-only": "npm run test:build && jest",
```
Comment (evilebottnawi) [inline package.json]: @michael-ciniawsky when i write code i want run only test to avoid regression (without lint), full test with lint and test i do before PR :smile:

File: lib/index.js (hunk 2 above)
Comment (evilebottnawi) [inline lib/index.js]: @michael-ciniawsky we don't load postcss config file when `config` not present or present but without `path`.

File: lib/index.js (hunk 1 above)
Comment (evilebottnawi) [inline lib/index.js]: @michael-ciniawsky weird, should works, I will test in the near future

File: lib/index.js [hunk 4]
```diff
@@ -47,78 +47,67 @@ module.exports = function loader (css, map) {
...
+    let resolvedOptions = Object.assign({
       to: file,
       from: file,
       map: sourceMap
         ? sourceMap === 'inline'
           ? { inline: true, annotation: false }
           : { inline: false, annotation: false }
         : false
-    }, config.options)
+    }, options, config.options)
```
Comment (evilebottnawi) [inline lib/index.js]: @michael-ciniawsky using `options` allow change `map`, it is bad, for some it is misleading behavior, but `to` and `from` now also we can change through `config.options`. I think about this.

File: lib/index.js (hunk 2 above)
Comment (michael-ciniawsky) [inline lib/index.js]: `postcss-load-config` has it's config lookup algo backed in via cosmiconfig, `options.config.path` should only be used for cases like 

```
|– .config
| |– postcss.config.js
|– src
| |– style.css
|
|– webpack.config.js
```

The current trigger for loading `postcss.config.js` is if no other options are set in `webpack.config.js` =>`postcss.config.js`

File: lib/index.js (hunk 4 above)
Comment (michael-ciniawsky) [inline lib/index.js]: > using options allow change map, it is bad, for some it is misleading behavior

? 🙃  You mean that currently one has to set `map` via `{ sourceMap: true || 'inline' }` (`webpack.config.js`) exclusively?

>  but to and from now also we can change through config.options

They shouldn't be changeable, since File I/O is handled by the webpack (Runner) 😛

File: lib/index.js (hunk 2 above)
Comment (evilebottnawi) [inline lib/index.js]: @michael-ciniawsky current behavior not works as expected we should fix this, see https://github.com/postcss/postcss-loader/pull/234#issue-229052711 (point 2).

Comment (evilebottnawi) [inline lib/index.js]: @michael-ciniawsky two solution:
1. Load `postcss` config always and them merge option from loader with config options (i think loader options more important than `postcss` config option, but ).
2. Load `postcss` config only when `config: true` or `config: { path: '/path/to/config'}` (best way, no problems with perf and understandable using)

**Comment by evilebottnawi** [comment]
@michael-ciniawsky we should finish and merge this PR asap, i spend about 5 hours for migrate to new postcss loader in most of my projects, in half the configuration is not loaded or source maps not working, current loading algorithm is misleading and incorrect.

My solution:
1. By default `config` is false and don't load postcss config.
2. If your have `config: true` or `config: { ctx : {...options}}` we loading config (also working good for nested config). Now `config: true` is disable by schema, but we can fix this.
3. If your have `config: { path: '/path/to/config' }` we loading config (absolutly).

This may seem small breaking changed, but it is bugfix, because now postcss config don't load in most use case.

File: lib/index.js [hunk 5]
```diff
@@ -47,53 +47,55 @@ module.exports = function loader (css, map) {
...
+    if (!config.options) config.options = {}
 
     if (config.file) this.addDependency(config.file)
 
+    let sourceMap = options.sourceMap || config.options.map
     let plugins = config.plugins || []
-    let options = Object.assign({
+
+    // Disable override `to` option
+    if (config.options.to) delete config.options.to
```
Comment (evilebottnawi) [inline lib/index.js]: Disable override important options from postcss `config`. it is break build.

File: lib/options.js [hunk 6]
```diff
@@ -11,12 +11,11 @@ module.exports = function parseOptions (params) {
   else if (Array.isArray(params.plugins)) plugins = params.plugins
   else plugins = params.plugins
 
-  const options = {}
-
-  if (typeof params !== 'undefined') {
-    options.parser = params.parser
-    options.syntax = params.syntax
-    options.stringifier = params.stringifier
+  const options = {
```
Comment (evilebottnawi) [inline lib/options.js]: We always have params as object, because we have `const options = loaderUtils.getOptions(this) || {}`

File: lib/index.js [hunk 7]
```diff
@@ -47,53 +47,55 @@ module.exports = function loader (css, map) {
...
-    } else if (options.config && !sourceMap && length > 1) {
-      return parseOptions.call(this, options)
-    } else if (!options.config && sourceMap && length > 1) {
-      return parseOptions.call(this, options)
-    } else if (options.config && sourceMap && length > 2) {
-      return parseOptions.call(this, options)
+    // Read options from options.config only when options.config is object
+    if (typeof options.config === 'object' &&
+      options.config !== null &&
+      Object.prototype.toString.call(options.config) === '[object Object]'
```
Comment (evilebottnawi) [inline lib/index.js]: Special for `config: true`

File: lib/index.js [hunk 8]
```diff
@@ -47,53 +47,55 @@ module.exports = function loader (css, map) {
...
     }
 
     return postcssrc(rc.ctx, rc.path, { argv: false })
   }).then((config) => {
     if (!config) config = {}
+    if (!config.options) config.options = {}
 
     if (config.file) this.addDependency(config.file)
 
+    let sourceMap = config.options.map || options.sourceMap
```
Comment (evilebottnawi) [inline lib/index.js]: When we load postcss config, she can return `map: undefined`, but in loader we have `sourceMap: true`, because first look on `config.options.map` then on `options.sourceMap`.

File: lib/options.js [hunk 9]
```diff
@@ -11,12 +11,11 @@ module.exports = function parseOptions (params) {
...
+    parser: params.parser,
+    syntax: params.syntax,
+    stringifier: params.stringifier,
+    map: params.sourceMap
```
Comment (evilebottnawi) [inline lib/options.js]: `postcss-load-config` return `map`, we should also return `map` from `parseOptions` (consistently and don't misleading for future options).

File: lib/index.js [hunk 10]
```diff
@@ -47,53 +47,60 @@ module.exports = function loader (css, map) {
...
+        rc.ctx.options = options.config.ctx
+      }
     }
-  }
 
-  const sourceMap = options.sourceMap
+    return postcssrc(rc.ctx, rc.path, { argv: false })
+      .then((config) => {
+        if (!config) config = {}
+        if (!config.options) config.options = {}
```
Comment (evilebottnawi) [inline lib/index.js]: Only `postcss-config-loader` can return undefined

File: lib/index.js [hunk 11]
```diff
@@ -47,53 +47,57 @@ module.exports = function loader (css, map) {
...
 
     return postcssrc(rc.ctx, rc.path, { argv: false })
-  }).then((config) => {
-    if (!config) config = {}
+      .then((config) => {
+        if (!config.options.map) config.options.map = options.sourceMap
 
+        return config
+      })
+  }).then((config) => {
```
Comment (evilebottnawi) [inline lib/index.js]: `config` is always object `parseOptions` return object and `postcss-load-config` also return object

**Comment by michael-ciniawsky** [comment]
I'm not in favour of changing the config loader behaviour to `{config: true }`. This neglects the reason for `postcss.config.js` which is simply to have **no** in `webpack.config.js` for most cases at all. I'm also concerned about merging options from (`webpack.config.js`) && (`postcss.config.js`). It should be a single source of truth and therefore an either or, I don't see the usecase for this, but if besides an edge case or personal perference one exists we can change it.

 >Always load config if your pass config.path, now it is not working as expected if i pass config.path and sourceMap: false my config file not loading -> options.config && !sourceMap && length > 1 (i want to load config and set sourceMap use loader options ), also sometimes i want setup some option from options loader.

`{ sourceMap: false }`  is basically unnessecary (default), just omit it, we need to fix the conditional checks for sure agreed upon that, but without breaking the current behaviour

> Some perf (we don't create rc object because it is not necessary if i use only options without options.config)

👍 
 
> Allow to passing options to postcss throw loader.options, now it is possibly only using postcss config.

What kind of options besides `parser || syntax || stringifier && plugins` ? The only two suitable options for map are `external` or `inline` as an annotation comment, which works via `sourceMap: true || 'inline'`, `from, to` are handled by webpack aswell

**Comment by evilebottnawi** [comment]
@michael-ciniawsky i wan't to use `parser` as option in `postcss-loader` and want to load postcss config. Also i wan't to override plugins, but use config for `parser`, `stringifier` and etc. and many other options.

```js
// TODO
    // Refactor
    if (!options.config && !sourceMap && length) {
      return parseOptions.call(this, options)
    } else if (options.config && !sourceMap && length > 1) {
      return parseOptions.call(this, options)
    } else if (!options.config && sourceMap && length > 1) {
      return parseOptions.call(this, options)
    } else if (options.config && sourceMap && length > 2) {
      return parseOptions.call(this, options)
    }
```

Really don't work in many my project, And I had to roll back (use `postcss-loader` v1). We need to make this explicit, if I want to use configurations (use `path` in `config` or use `true`) then use them. 

This changes works on all my project, testing with `create-react-app`, `angular-starter` and some other popular boilerplate. Btw, current algorithm don't work in `create-react-app` (just disable source map and put config). Seriously it just does not work.

If you do not intend to change the logic of work to explicit, then probably I should do fork, because many in many of my project it is not working as expected, also where i use `create-react-app` it too don't work. It is critical for me.

**Comment by evilebottnawi** [comment]
`{ sourceMap: false } is basically unnessecary (default), just omit it, we need to fix the conditional checks for sure agreed upon that, but without breaking the current behaviour`, i can't change this in many boilerplate, because it is using for all loader from config.

config.js
```
module.exports = {
    ...
    sourceMap: false
    ...
};
```
And in `webpack.config.js` it is setup to all loaders.

**Comment by michael-ciniawsky** [comment]
The current loading algo is crap for sure, but still the behaviour shouldn't change :)

❌ `{ config: true }` or similiar isn't an option (neglect's zero config approach)

**postcss.config.js**
```js
module.exports = ({ file, options, env }) => ({
   // Maybe support options for `map` (PostCSS),
   // but `true` || `inline` handles that fine for the 99% cases,
   // webpack also sets specific requirements here
    sourceMap: options.sourceMap 
});
```

The reason for not supporting `sourceMap` within  `postcss.config.js` is during the fact that, it should work with different PostCSS middlewares and therefore only supports PostCSS options && plugins (middleware agnostic). All other middleware related options are set, where the particular middleware normally provides an interface to do so. `{ sourceMap: true }` is a webpack (`postcss-loader`) specific option and is better defined in `webpack.config.js`

**webpack.config.js**
```js
{
   test: '..',
   use: [
     { loader: 'style-loader', options: { sourceMap: env === 'development' ? true : false} },
     { loader: 'css-loader', options: { sourceMap: env === 'development' ? true : false } },
     // { loader: 'postcss-loader', options: { sourceMap: env === 'development' ? true : false } },
     // Usage worst case scenario
     { 
       loader: 'postcss-loader', 
       options: { config: { ctx: { sourceMap: env === 'development' ? true : false } } } 
     },
     { loader: 'sass-loader', options: { sourceMap: env === 'development' ? true : false} }
   ]
}
```

All we need is to find a better way to filter the loader options and either parse the options, if we have additional ones (`parser &&|| plugins`), or try to load a config file

**Comment by evilebottnawi** [comment]
@michael-ciniawsky Today I will create a table with all the options and behavior that should, so it will be easier to come to a final decision.

**Comment by evilebottnawi** [comment]
@TrySound Maybe you can help us and with this?

**Comment by evilebottnawi** [comment]
@michael-ciniawsky now using `css` inside `sass/less/etc` throw error, https://github.com/insin/nwb/issues/311#issuecomment-304957316 (https://webpack.slack.com/archives/C3NUYSYF4/p1496168402383007)

**Comment by evilebottnawi** [comment]
@michael-ciniawsky also adding any options to `options` (`test: true`) break all loading logic.

**Comment by michael-ciniawsky** [comment]
@evilebottnawi What is the issue in insin/nwb#311 (comment) regarding `postcss-loader`?

> also adding any options to options (test: true) break all loading logic.

Any additional properties should ideally throw `OptionsValidationError`, what is the propose of `{ test: true }` or is the a stub example?

**Comment by evilebottnawi** [comment]
@michael-ciniawsky now it is not throw `OptionsValidationError` :disappointed:  and break loading logic. But is simple
```sass
@import "~swiper/dist/css/swiper.css";
```
in `style-loader` -> `css-loader` -> `postcss-loader` -> `sass-loader`

**Comment by michael-ciniawsky** [comment]
I don't get the issue with the example ? Yeah.. `additionalProperties` is set to true, bc `plugins` prop `{Function}` isn't testeable with `schema-utils` atm. Still additional properties should fail, but with `ValidationError`

**Comment by evilebottnawi** [comment]
@michael-ciniawsky 
```js
webpackConfig.module.rules.push({
        test: /\.s[ac]ss$/,
        use: [
            {
                loader: 'style-loader',
                options: {
                    sourceMap: true
                }
            },
            {
            loader: 'css-loader',
            options: {
                import: true,
                sourceMap: true,
                importLoaders: 2
            }
        },
        {
            loader: 'postcss-loader',
            options: {
                sourceMap: true,
                plugins: () => [require('autoprefixer')]
            }
        },
        {
            loader: 'sass-loader',
            options: { 
                sourceMap: true
            }
        }
    });
```

**Comment by michael-ciniawsky** [comment]
And were does it exactly choke, seems unrelated to `postcss-loader`

**Comment by evilebottnawi** [comment]
@michael-ciniawsky `Module build failed: Error: No PostCSS Config found in:` because tries to load the configuration and `css-loader` (https://github.com/webpack-contrib/css-loader/blob/master/lib/getImportPrefix.js#L12) return only `{ sourceMap: true }`, seems not respect function.

Problems:
1. We need disallow to using `plugins` as function or try to solve respect function option in https://github.com/webpack-contrib/css-loader/blob/master/lib/getImportPrefix.js#L12
2. Throw `OptionsValidationError` on additional argument in options (not a few people are setting the wrong options in an unnecessary place).
3. Also some strange behavior: using `importLoaders` for handling `@import` i got options in `postcss-loader` from `style-loader` -> `css->loader` -> `postcss-loader` (**actual options**) -> `sass-loader` (`test: /\.s[ac]ss$/,`), but i have difference plugins for `sass` and pure `css` (maybe expected behavior), but i was surprised, i think it is using `style-loader` -> `css->loader` -> `postcss-loader` (**expected options**) (`test: /\.css$/`)

**Comment by evilebottnawi** [comment]
@michael-ciniawsky why we use `parseOptions` then:
1. `options.config && !sourceMap && length > 1`
2. `options.config && sourceMap && length > 2`

Can i see example configuration for this use case?

**Comment by evilebottnawi** [comment]
@ai Maybe you can help us solve this problem :disappointed:

**Comment by michael-ciniawsky** [comment]
The `ident` seems to be the problem again

```diff
ExtractTextPlugin.extract({
            fallback: 'style-loader',
            use: [
              {
                loader: 'css-loader',
                options: { importLoaders: 2, sourceMap: true }
              },
              {
                loader: 'postcss-loader',
                options: {
                  config: { ctx: { test: 1 } },
   -               ident: 'postcss',
                  plugins: [ require('postcss-nested')() ],
                  sourceMap: true
                }
              },
              {
                loader: 'sass-loader', options: { sourceMap: true }
              }
            ]
})
```

```
    ERROR in ../~/css-loader?{"importLoaders":2,"sourceMap":true}!../~/postcss-loader/lib?{"config":{"ctx":{"test":1}},"plugins":[null],"sourceMap":true}!../~/sass-loader/lib/loader.js?{"sourceMap":true}!./components/component/index.css
    Module build failed: TypeError: Cannot read property 'postcss' of null
        at Processor.normalize (/Users/Cini/Github/Webpack/__test__/node_modules/postcss-loader/node_modules/postcss/lib/processor.js:136:12)
        at new Processor (/Users/Cini/Github/Webpack/__test__/node_modules/postcss-loader/node_modules/postcss/lib/processor.js:51:25)
        at postcss (/Users/Cini/Github/Webpack/__test__/node_modules/postcss-loader/node_modules/postcss/lib/postcss.js:73:10)
        at Promise.resolve.then.then (/Users/Cini/Github/Webpack/__test__/node_modules/postcss-loader/lib/index.js:137:12)
        at <anonymous>
     @ ../~/css-loader?{"importLoaders":2,"sourceMap":true}!../~/postcss-loader/lib?{"config":{"ctx":{"test":1}},"plugins":[null],"sourceMap":true}!../~/sass-loader/lib/loader.js?{"sourceMap":true}!./App.scss 3:10-313
```

```diff
ExtractTextPlugin.extract({
            fallback: 'style-loader',
            use: [
              {
                loader: 'css-loader',
                options: { importLoaders: 2, sourceMap: true }
              },
              {
                loader: 'postcss-loader',
                options: {
                  config: { ctx: { test: 1 } },
 +               ident: 'postcss',
                  plugins: [ require('postcss-nested')() ],
                  sourceMap: true
                }
              },
              {
                loader: 'sass-loader', options: { sourceMap: true }
              }
            ]
})
```

```
Hash: 730caef604ad1e57d3b2
Version: webpack 2.6.1
Time: 4578ms
        Asset       Size  Chunks             Chunk Names
     index.js  683 bytes   index  [emitted]  index
    vendor.js    5.98 kB  vendor  [emitted]  vendor
    index.css  324 bytes   index  [emitted]  index
 index.js.map  938 bytes   index  [emitted]  index
index.css.map  846 bytes   index  [emitted]  index
vendor.js.map    6.07 kB  vendor  [emitted]  vendor
[./App.js] ./App.js 97 bytes {index} [built]
[./App.scss] ./App.scss 41 bytes {index} [built]
Child extract-text-webpack-plugin:
    [../node_modules/css-loader/index.js?{"importLoaders":2,"sourceMap":true}!../node_modules/postcss-loader/lib/index.js?{"config":{"ctx":{"test":1}},"ident":"postcss","plugins":[null],"sourceMap":true}!../node_modules/sass-loader/lib/loader.js?{"sourceMap":true}!./App.scss] ../~/css-loader?{"importLoaders":2,"sourceMap":true}!../~/postcss-loader/lib?{"config":{"ctx":{"test":1}},"ident":"postcss","plugins":[null],"sourceMap":true}!../~/sass-loader/lib/loader.js?{"sourceMap":true}!./App.scss 1.36 kB {0} [built]
    [../node_modules/css-loader/lib/css-base.js] ../~/css-loader/lib/css-base.js 2.26 kB {0} [built]
        + 1 hidden modules
```

**Comment by michael-ciniawsky** [comment]
The same goes for the `importLoaders` issue, it works with `postcss.config.js`, but chokes on `options.plugins` atm, needs more investigation

**Comment by evilebottnawi** [comment]
@michael-ciniawsky ok then current implementation is correct, only need throw

**Comment by michael-ciniawsky** [comment]
@evilebottnawi Are you intending to update this PR or can I close and open a new one ? 😛

**Comment by evilebottnawi** [comment]
@michael-ciniawsky What is wrong with this PR now? Check on only `plugins` and `config` is incorrectly, i can `use` only `parser` and `stringifier` for parsing and output in other style.

**Comment by michael-ciniawsky** [comment]
`sourceMap`  should be the only way to steup source maps
 `postcss.config.js` is still not **implicitly** loaded

**Comment by evilebottnawi** [comment]
@michael-ciniawsky i.e. we ignore source map from `postcss.config.js`?
It would be nice to see the options under which the config should not boot

**Comment by michael-ciniawsky** [comment]
The sourcemap is mainly because of the loader chain 

```js
[
   { loader: 'style-loader', options: { sourceMap: true } },
   { loader: 'css-loader', options: { sourceMap: true } },
   'postcss-loader' ?sourceMap
   { loader: 'style-loader', options: { sourceMap: true } }
]
```

Also webpack user will then need to know, **how** to config PostCSS Sourcemap Options  instead of `true || 'inline'`

**Comment by evilebottnawi** [comment]
@michael-ciniawsky we already have `inline` or `true` source map. We ignore the source maps from `postcss.config.js`?

**Comment by michael-ciniawsky** [comment]
> we already have inline or true source map. We ignore the source maps from postcss.config.js?

☝️ https://github.com/postcss/postcss-loader/pull/234#issuecomment-308409814

The missing `options.exec` check and #250 + maybe throw if `options.plugins && options.config` are set at the same time, is all we can do in the `v2.0.0` line, but better with a fresh PR :)

**Comment by evilebottnawi** [comment]
@michael-ciniawsky why your close? I will now have to do most of it again, having spent this time, thanks for respecting my time

**Comment by evilebottnawi** [comment]
Main of problems, we have ugly algo and few test, I just do not understand when we need to load config when not, and no specific comments, I'm glad to help and close this problem, but I just do not know what's wrong now :disappointed:

**Comment by evilebottnawi** [comment]
@michael-ciniawsky 
1. Please describe all options when we load config when not or add tests, now behavior is misleading and no one understands except you.
2. What wrong with source map now? We load source map from options, if your have config and `map` we get source map from config, if your have config but don't have `map` we get they from loader `options`?

**Comment by michael-ciniawsky** [comment]
kk mom I'm trying to sum it up

**Comment by evilebottnawi** [comment]
@michael-ciniawsky Please understand me, we have an internal boilerplate config where we can't remove `sourceMap` from source code, it is global variable for all loaders and we need support `postcss.config.js`, it is bug really critical (seems other people have same problem), we can't fix this along time, let's do it **asap**

**Comment by evilebottnawi** [comment]
Let's do it:
1. If we have `exec` and/or `parser` and/or `syntax` and/or `stringifier` and/or `plugins` - we don't load config.
2. If we have `config` - we load config
3. If we don't have options - we load config.
4. sourceMap option does not affect logic, we can use `sourceMap` in 1 and 2 and 3 point.

**Comment by evilebottnawi** [comment]
@TrySound can your review my commit? All tests fine

**Comment by michael-ciniawsky** [comment]
@evilebottnawi Please reduce the PR here only to your proposed `sourceMap` changes

**Comment by evilebottnawi** [comment]
@michael-ciniawsky this PR contain only relevant changed

File: lib/index.js [hunk 12]
```diff
@@ -47,53 +47,55 @@ module.exports = function loader (css, map) {
...
-      },
-      options: {}
-    }
-  }
-
-  if (options.config) {
-    if (options.config.path) {
-      rc.path = path.resolve(options.config.path)
+  Promise.resolve().then(() => {
+    if (options.exec || options.parser || options.syntax || options.stringifier || options.plugins) {
```
Comment (michael-ciniawsky) [inline lib/index.js]: rm

File: lib/index.js [hunk 13]
```diff
@@ -47,53 +47,55 @@ module.exports = function loader (css, map) {
...
+      return parseOptions.call(this, options)
     }
 
-    if (options.config.ctx) {
-      rc.ctx.options = options.config.ctx
+    const rc = {
```
Comment (michael-ciniawsky) [inline lib/index.js]: rm

File: package.json [hunk 14]
```diff
@@ -32,6 +32,7 @@
     "docs": "jsdoc2md lib/index.js > LOADER.md",
     "pretest": "npm run lint && npm run test:build",
     "test": "jest",
+    "test-only": "npm run test:build && jest && npm run clean",
```
Comment (michael-ciniawsky) [inline package.json]: rm

File: lib/index.js [hunk 15]
```diff
@@ -47,53 +47,55 @@ module.exports = function loader (css, map) {
...
-    let options = Object.assign({
+
+    // Disable override `to` option
+    if (config.options.to) delete config.options.to
+    // Disable override `from` option
+    if (config.options.from) delete config.options.from
+    // Disable override `map` option
+    if (config.options.map) delete config.options.map
+
+    let postcssOption = Object.assign({
```
Comment (michael-ciniawsky) [inline lib/index.js]: Why was this renamed, should be `options`

File: lib/index.js [hunk 16]
```diff
@@ -47,53 +47,55 @@ module.exports = function loader (css, map) {
...
-      return parseOptions.call(this, options)
+      if (options.config.ctx) {
+        rc.ctx.options = options.config.ctx
+      }
     }
 
     return postcssrc(rc.ctx, rc.path, { argv: false })
-  }).then((config) => {
-    if (!config) config = {}
+      .then((config) => {
```
Comment (michael-ciniawsky) [inline lib/index.js]: wrong `ident`

**Comment by evilebottnawi** [comment]
@michael-ciniawsky `true` || `inline`

File: lib/index.js (hunk 15 above)
Comment (evilebottnawi) [inline lib/index.js]: @michael-ciniawsky because `options` related to `loader` (`const options = loaderUtils.getOptions(this) || {}`), `postcssOptions` related to `postcss` options

File: package.json (hunk 14 above)
Comment (evilebottnawi) [inline package.json]: @michael-ciniawsky How can I run only tests?

File: lib/index.js (hunk 16 above)
Comment (evilebottnawi) [inline lib/index.js]: @michael-ciniawsky don't understand, please provide example

File: lib/index.js (hunk 13 above)
Comment (evilebottnawi) [inline lib/index.js]: @michael-ciniawsky we don't need rc if we don't load config

File: lib/index.js (hunk 12 above)
Comment (evilebottnawi) [inline lib/index.js]: @michael-ciniawsky why? your ugly algo is don't work as expected

**Comment by michael-ciniawsky** [comment]
But that's not the only two options for `map`, what happens of e.g `{ inline: true, annotation: false }` is directly passed via `postcss.config` does it work ?

**Comment by evilebottnawi** [comment]
@michael-ciniawsky In your code https://github.com/postcss/postcss-loader/blob/master/lib/index.js#L99 `true` and `inline` only work. Now algo is same.

File: lib/index.js (hunk 13 above)
Comment (michael-ciniawsky) [inline lib/index.js]: #259

File: lib/index.js (hunk 12 above)
Comment (michael-ciniawsky) [inline lib/index.js]: #259

File: lib/index.js (hunk 16 above)
Comment (michael-ciniawsky) [inline lib/index.js]: the code indentation is off :), at least looks it

**Comment by evilebottnawi** [comment]
@michael-ciniawsky Do you have specific examples of why this code is not working (example of config or failed tests), all tests are passed, or is this a subjective non-constructive mono solution?

File: lib/index.js (hunk 15 above)
Comment (michael-ciniawsky) [inline lib/index.js]: It works atm due to lexical scope, as long as there is no need to rename it please leave it as is

File: lib/index.js (hunk 16 above)
Comment (evilebottnawi) [inline lib/index.js]: @michael-ciniawsky 
```js
return postcssrc(rc.ctx, rc.path, { argv: false })
      .then((config) => {
        // Get `sourceMap` option from loader options when no `map` option in `postcss.config.js`
        if (!config.options.map) config.options.map = options.sourceMap

        return config
      })
```
Where?

**Comment by michael-ciniawsky** [comment]
> Do you have specific examples of why this code is not working (example of config or failed tests), all tests are passed, or is this a subjective non-constructive mono solution?

You mean the config resolving here ?

**Comment by evilebottnawi** [comment]
@michael-ciniawsky i mead where algo don't work as expected

**Comment by evilebottnawi** [comment]
@michael-ciniawsky https://github.com/postcss/postcss-loader/pull/259 Thanks for respect my work and my time.

**Comment by michael-ciniawsky** [comment]
Why are you closing this now? 🙃

**Comment by evilebottnawi** [comment]
@michael-ciniawsky because I'm tired of non constructive argument and ignoring, for me it will be faster to create fork. 

Because your say:
> The only two suitable options for map are external or inline as an annotation comment, which works via sourceMap: true || 'inline', from, to are handled by webpack aswell

**Comment by michael-ciniawsky** [comment]
What do you mean by non-constructive argument and ignoring ? Ignoring is just plain wrong... `Disagreement !== Ignoring`

Constrcutive or not is subjective, but

- The PR at first changed the config loading **enterily** and I explained more then once, why I don't think it is a good idea to do it that way 

- For `sourceMap` my concern is that there will be confusion about it, when `postcss.config.js` supports it, but I'm not against it in general (there would still edge case in the PR **as is** )

- The other PR was not meant offensive in any way and I have had no problem by just closing it, I don't care about these things


```js
let options = Object.assign({
      to: file,
      from: file,
      map: sourceMap // <=
        ? sourceMap === 'inline'
          ? { inline: true, annotation: false }
          : { inline: false, annotation: false }
        : false
}, config.options)
```

This would require additional changes to work with **all** possible options `map` PostCSS (`postcss.config.js`) supports

**Comment by evilebottnawi** [comment]
@michael-ciniawsky 
1. your just close my PR without explaining what is wrong - it is non-constructive.
2. Your say 
> The only two suitable options for map are external or inline as an annotation comment, which works via sourceMap: true || 'inline', from, to are handled by webpack aswell
Because i do only `true` and `inline` value supporting for sourceMap
3. `Ignoring` - https://github.com/postcss/postcss-loader/pull/234#issuecomment-303085769, https://github.com/postcss/postcss-loader/pull/234#issuecomment-304947846, https://github.com/postcss/postcss-loader/pull/234#issuecomment-305237199, https://github.com/postcss/postcss-loader/pull/234#issuecomment-306444817
4. No Respect for my time and effort, thanks

 Just don't ignore https://github.com/postcss/postcss-loader/pull/259 and merge PR to solve issue

**Comment by michael-ciniawsky** [comment]
kk, can we please in general agree to take it a bit more relaxed and avoid the childish behaviour ?

> your just close my PR without explaining what is wrong - it is non-constructive.

When I closed the PR is still changed the loading behaviour, that was the reason

> Your say

> The only two suitable options for map are external or inline as an annotation comment, which 
> works via sourceMap: true || 'inline', from, to are handled by webpack aswell
> Because I do only true and inline value supporting for sourceMap


Yes, that's the two ways to handle sourcemaps, `postcss.config.js` supports `{Object}` e.g `{ inline: true, annotation: false }` which `postcss-loader` needs handle aswell then, which is ok if needed

> Ignoring - #234 (comment), #234 (comment), #234 (comment), #234 (comment)

I answer all of them and the second isn't even addressed at me 🙃 

> No Respect for my time and effort, thanks

I hearing this for the third+ time now, what about other people's time ? This is just trying to personify matters to passive aggressively offend  && push others to do what you want?  As I said I don't mind closing my PR's, waste my time in your speaking and in general the least thing I intend to do is to actively disrespect others. Which is disagree I did in anyway, but if you felt that way I apologize

**Comment by evilebottnawi** [comment]
@michael-ciniawsky thanks for apologize, let's first fix unexpected behavior with sourceMap https://github.com/postcss/postcss-loader/pull/259

**Comment by evilebottnawi** [comment]
@michael-ciniawsky and then solve problems: should we respect map from `postcss.config.js` and how we should do this, or now all ok

**Comment by michael-ciniawsky** [comment]
Yes, but we need to add a check / refactor https://github.com/postcss/postcss-loader/blob/master/lib/index.js#L99 properly to support 'full' PostCSS Sourcemap Options `{Object}`, also the `config.options.from/comfig.options.to`  sanitizing you added was a good idea

**Comment by michael-ciniawsky** [comment]
Also check if some possible PostCSS Sourcemap Options doesn't play well with webpack (I don't know 💯 yet)
